### DISCUSSION TITLE
Capacitor 3 Upgrade: Updating Android packages in Project Structure Suggestions

### DISCUSSION THREAD

**Discussion by KevinKelchen** [ionic-team/capacitor#5101] [Q&A] [answered]
Hi there! 👋

I am currently upgrading our app from Capacitor 2 to Capacitor 3 and had a question about a [suggestion](https://capacitorjs.com/docs/updating/3-0#update-gradle-to-70) in the upgrade guide.

First of all, I want to give a BIG "THANK YOU!" to those who determined the scope of Capacitor 3 and those who worked on the upgrade guide! Upgrading from v2 to v3 has been fairly straightforward because most of the changes involved the splitting-out of plugins without lumping a lot of other changes into the mix. The smoothness of the upgrade is also notably attributed to the detailed and well-written upgrade guide. That guide has been very valuable--especially as it pertains to maintaining two native projects with many project files. Trying to diff our existing files with the files from the current Android and iOS project templates would be quite tedious. Bravo! 👏

The upgrade guide [says](https://capacitorjs.com/docs/updating/3-0#update-gradle-to-70), "You may want to evaluate suggested updates to Android packages in the Suggestions section of the Project Structure dialog." My suggested updates look like thus:

![image](https://user-images.githubusercontent.com/8027502/136080119-20932e48-0342-4aa7-aead-2872493249b0.png)

My question is: How do I know which updates are safe to take? I know that depends on each project, but are there any general guidelines?

What I'm afraid of messing up is some specific configuration needed for Cordova plugins in our project. I can check out the projects for all such plugins and try to make a determination but thought I'd ask here for any general recommendations.

Thanks so much! 😀

Kevin

**Comment by jcesarmobile** [comment, marked as answer]
Over time google has proven that they don't follow semver on their libraries, they have introduced breaking changes in minor or patch version bumps. (maybe they didn't do it on purpose)

But I would recommend to always use latest (stable) available, and if you find any issue, downgrade to the previous one or until everything works again.

In the upgrade guides we put the version that is latest at the moment of writing the guide and verify that everything works, so that's why we recommend those.

I've noticed that you have two dependencies for android support libraries (v4), one on your app and another one on a cordova plugin.
You shouldn't need the one in your app anymore since Capacitor switched to AndroidX dependencies in Capacitor 2 and the old android support libraries are not compatible.
For the cordova plugin, the plugin should be updated to not use android support libraries too now that cordova-android also uses AndroidX dependencies.

For the cordova-android dependency, I've tested and you can't use version 10 until we merge and release [this PR](https://github.com/ionic-team/capacitor/pull/5103), but using cordova-android 10 requires to bump your minSDK to 22 instead of 21.

**Reply by KevinKelchen** [reply]
Hi Julio, 😀

Thank you for the thoughtful and kind response! I really appreciate you taking the time to help answer my questions! 😊

> Over time google has proven that they don't follow semver on their libraries, they have introduced breaking changes in minor or patch version bumps. (maybe they didn't do it on purpose)
> 
> But I would recommend to always use latest (stable) available, and if you find any issue, downgrade to the previous one or until everything works again.
> 
> In the upgrade guides we put the version that is latest at the moment of writing the guide and verify that everything works, so that's why we recommend those.

Thanks for the info here! That all makes sense to me.


> I've noticed that you have two dependencies for android support libraries (v4), one on your app and another one on a cordova plugin. You shouldn't need the one in your app anymore since Capacitor switched to AndroidX dependencies in Capacitor 2 and the old android support libraries are not compatible. For the cordova plugin, the plugin should be updated to not use android support libraries too now that cordova-android also uses AndroidX dependencies.

That is good to know. I will try removing the Android Support Library dependency from the `app` module (as opposed to clicking `Update`).

Regarding the `capacitor-cordova-android-plugins` module's dependency on the Android Support Library, I may need to retain that due to a couple of Cordova plugins we use that have not migrated to AndroidX ([cordova-plugin-file-opener2](https://github.com/pwlin/cordova-plugin-file-opener2/issues/256#issuecomment-657574795) and [phonegap-plugin-barcodescanner](https://github.com/phonegap/phonegap-plugin-barcodescanner/issues/878)). We use `jetify` for at least those dependencies. Perhaps in that case I need to retain the `capacitor-cordova-android-plugins` dependency on the Android Support Library, but perhaps I can update from `4:27.+` to `28.0.0` although I'd probably have to see what breaking changes there are and account for what you said about Google and semver adherence.


> For the cordova-android dependency, I've tested and you can't use version 10 until we merge and release [this PR](https://github.com/ionic-team/capacitor/pull/5103), but using cordova-android 10 requires to bump your minSDK to 22 instead of 21.

Also very good to know! At least from a `minSdkVersion` standpoint we're set to `24` so that shouldn't be a blocker; we'd probably just need to wait on the new Capacitor release.


So here's a funny thing: to at least attempt to upgrade _something_ I tried updating the following by clicking `Update Variable`:
![image](https://user-images.githubusercontent.com/8027502/136257047-91ed86c0-e4c5-466e-a745-1e931174973a.png)

With managing these dependencies being new to me, I also wasn't sure whether to choose `Update Variable` or `Update Dependency`. 😅 At any rate, after clicking `Update Variable` I then got a lot more suggested Updates:

![image](https://user-images.githubusercontent.com/8027502/136257301-12f13b63-9dd7-450b-9cb5-bc6e7430db5a.png)

![image](https://user-images.githubusercontent.com/8027502/136257342-e696c4e9-0db0-499b-ac77-aa332e0e3767.png)

![image](https://user-images.githubusercontent.com/8027502/136257360-ff1d792a-15a2-4bb3-b4c7-0651ae7d53bd.png)

![image](https://user-images.githubusercontent.com/8027502/136257381-17266201-067f-4785-83d3-3afc7d106fd8.png)


Sorry, I know that's kind of overwhelming. 😅 A lot of those have similar suggested updates so perhaps that will help make the decision-making easier, but I'm interested in any thoughts you might have!

Is it also expected that doing what I did--clicking `Update Variable` for something--would result in more suggested Updates?


Thank you so much! 😀

Kevin

**Reply by KevinKelchen** [reply]
> I will try removing the Android Support Library dependency from the `app` module (as opposed to clicking `Update`).

Heh, when I try to remove the dependency and click `Apply`, it comes back. 😄

I then tried first updating it to `28.0.0` and then attempt dependency removal but it still pops back in there.

![image](https://user-images.githubusercontent.com/8027502/136260614-dd579d78-ebc1-4b4b-ae90-2af91b73aeee.png)

Wonder if that has something to do with the `capacitor-cordova-android-plugins` module having the dependency as well and since `app` depends on `capacitor-cordova-android-plugins` the Android Support Library is a transitive dependency of `app` until `capacitor-cordova-android-plugins` has Android Support Library removed.

Also, I've noticed after closing and going back into the Android Studio `Project Structure` dialog that the deluge of suggested updates no longer appear. 🤔 Maybe I need to change something while the dialog is open to get the many additional suggestions to return. Not sure what the normal behavior is.

**Reply by jcesarmobile** [reply]
Use update variable as it updates the value in `variables.gradle`, update dependency replaces the existing variable usage with the numeric value, so you stop using variables and can cause version mismatch.

At first do it only for your app, not for plugins or other modules, after apply quit that screen and do a gradle sync (elephant button), since all plugins have the same variables your app have, when you go back to the suggestions screen I don't think it will show the suggestions anymore since they should be picking the new variable value.

`capacitor-cordova-android-plugins` content (including gradle files) gets regenerated on every `npx cap update/sync`, so don't do any suggested changes there.
I've checked both cordova plugins and both of them use `$ANDROID_SUPPORT_V4_VERSION` variable, so you could add `ANDROID_SUPPORT_V4_VERSION` variable  (without the $) in your `variables.gradle` and set the value to 28 there and then on sync/update it won't use 27+ but keep the variable.
But anyway, since you are using jetifier, it will replace the android support v4 with the AndroidX equivalent, so don't worry too much about it.

**Reply by KevinKelchen** [reply]
This is exactly what I needed to know! ❤️

Thank you so much, Julio! 😀 Your answers have been superb and your help has been super valuable!

Take care,

Kevin
